### PR TITLE
8341243: Use ArraySupport.SOFT_MAX_ARRAY_LENGTH for max array size in java.base

### PR DESCRIPTION
**Pull request by eirbjo**
Please review this cleanup PR which updates code and tests in `java.base` to consistently use  `jdk.internal.util.ArraySupport.SOFT_MAX_ARRAY_LENGTH` when referring to the JVM's maximum array size implementation limit. Currently, instances of `Integer.MAX_VALUE - 8` are found across the code base, with varying degrees of documentation. It would be good to consolidate on a single source of truth, with proper documentation.

This PR is a follow-up to #20905 where the same change was requested in `java.util.zip`.

My understanding is that javac will fold this constant value into the byte code of the compiled use sites, as such this change should not affect class loading or cause startup issues. 

Instances selected for this PR were found searching for "Integer.MAX_VALUE - 8". The PR replaces these with `ArraySupport.SOFT_MAX_ARRAY_LENGTH`, while trimming or amending some code comments where appropriate. (`SOFT_MAX_ARRAY_LENGTH` already has a good explainer which does not need repetition at each use site).

I also searched for instances of `Integer.MAX_VALUE - 1` and `Integer.MAX_VALUE - 2`, no convincing candidates were found.

Instances outside `java.base` were deliberately left out to limit the scope and review cost of this PR.

Tests updated to use `SOFT_MAX_ARRAY_LENGTH` are updated with the jtreg tag `@modules java.base/jdk.internal.util`.

Testing: No new tests are added in this PR, the `noreg-cleanup` label is added to the JBS. The five affected tests have been run manually. GHA tests run green.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8341243](https://bugs.openjdk.org/browse/JDK-8341243): Use ArraySupport.SOFT_MAX_ARRAY_LENGTH for max array size in java.base (**Enhancement** - P4)

### Reviewers
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)
 * [Stuart Marks](https://openjdk.org/census#smarks) (@stuart-marks - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21268/head:pull/21268` \
`$ git checkout pull/21268`

Update a local copy of the PR: \
`$ git checkout pull/21268` \
`$ git pull https://git.openjdk.org/jdk.git pull/21268/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21268`

View PR using the GUI difftool: \
`$ git pr show -t 21268`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21268.diff">https://git.openjdk.org/jdk/pull/21268.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21268#issuecomment-2383968473)

### CODE REVIEW

File: test/jdk/java/lang/StringBuilder/HugeCapacity.java [hunk 1]
```diff
@@ -75,7 +78,9 @@ private static void testUtf16() {
 
     private static void testHugeInitialString() {
         try {
-            String str = "Z".repeat(Integer.MAX_VALUE - 8);
+            // Maximum array length supported by the JVM implementation
+            int maxArrayLength = ArraysSupport.SOFT_MAX_ARRAY_LENGTH;
+            String str = "Z".repeat(maxArrayLength);
```
Comment (stuart-marks) [inline test/jdk/java/lang/StringBuilder/HugeCapacity.java:83]: The comment is a bit misleading. I'm not sure any comment is warranted, as you had observed, there's a big description next to SOFT_MAX_ARRAY_LENGTH. I'm also not sure a new local variable maxArrayLength adds anything.

File: test/jdk/java/lang/StringBuffer/HugeCapacity.java [hunk 2]
```diff
@@ -43,7 +46,9 @@ public static void main(String[] args) {
 
     private static void testHugeInitialString() {
         try {
-            String str = "Z".repeat(Integer.MAX_VALUE - 8);
+            // Maximum array length supported by the JVM implementation
+            int maxArrayLength = ArraysSupport.SOFT_MAX_ARRAY_LENGTH;
+            String str = "Z".repeat(maxArrayLength);
```
Comment (stuart-marks) [inline test/jdk/java/lang/StringBuffer/HugeCapacity.java:51]: Same comment as in the corresponding code in the StringBuilder test.

**Comment by stuart-marks** [comment]
As this stands (modulo my other comments) this change is mostly OK. Using the SOFT_MAX value within java.base is fine. Using SOFT_MAX within java.base-related tests is a little suspicious, because it requires the addition of directives that export an internal package from java.base ... but these tests are generally expected to be closely coupled with java.base internals, so they're probably ok too.

I agree with the decision not to expand usage of SOFT_MAX further, as it's really just a guess at a VM implementation limit that's not defined anywhere. (There have been requests to make this public; see [JDK-8246725](https://bugs.openjdk.org/browse/JDK-8246725), which I'm opposed to.) The *real* fix here is to fix the JVM so that it can allocate arrays of Integer.MAX_VALUE. From time to time I hear rumors that this might happen at some point. Meanwhile, let's not expose something with ill-defined semantics like SOFT_MAX beyond java.base.

As an aside, for your own edification, you might want to verify that compile-time constants are indeed inlined, by inspecting the before-and-after bytecode. This is well documented; see JLS 13.1, item 3 of the first numbered list in that section. I'm sure this is well tested, so there's no need for a test for this.

File: test/jdk/java/lang/StringBuilder/HugeCapacity.java (hunk 1 above)
Comment (eirbjo) [reply test/jdk/java/lang/StringBuilder/HugeCapacity.java]: > The comment is a bit misleading. I'm not sure any comment is warranted, as you had observed, there's a big description next to SOFT_MAX_ARRAY_LENGTH. I'm also not sure a new local variable maxArrayLength adds anything.

Thanks, I have removed the comment and variable for these two tests.

I guess I felt SOFT_MAX_ARRAY_LENGTH looked a bit lonely here, given that it was just used as an argument and had no constant or comment like in most other instances. But I agree that the comment and the variable does not add much value here, the name of the test makes it clear that this is about capacity and the documentation on SOFT_MAX_ARRAY_LENGTH is just a click away.

**Comment by stuart-marks** [comment]
@eirbjo Thanks for your thoughts on using the constant in the tests. It all makes good sense.

@jaikiran Thanks for handling testing for this.

**Review by stuart-marks — APPROVED** [review]

**Comment by eirbjo** [comment]
Thanks for taking time to review this Stuart. 

I'm delegating the integration of this PR such that @jaikiran can take care of that once his test run is complete.

**Comment by jaikiran** [comment]
tier1, tier2 and tier3 testing in our CI completed without any issues. I'll now go ahead with integrating this.

**Pull request closed without merging**